### ISSUE TITLE
Revisit `<FileTree>` accessibility

### ISSUE THREAD

**Issue by delucis** [withastro/docs#2722]
We use a `<FileTree>` component to generate pretty file trees from nested unordered lists. For example:

| Input | Output |
|---|---|
| ![image](https://user-images.githubusercontent.com/357379/220926559-5eb5c257-5cbb-48a2-b990-152fdb93bcc5.png) | ![image](https://user-images.githubusercontent.com/357379/220926744-a9b492ee-4f14-4859-bb9e-4b4c21cbc437.png) |

This improved on our previous use of codeblocks with ASCII art but might still not be as accessible as we can make it. VoiceOver for example doesn’t make it super obvious what list depth you are at so it’s hard to know the full path/position of a given file.

It would be good to revisit some details around this component to see what we can improve.

- Investigate whether [the Tree View Pattern documented in the ARIA Authoring Practices Guide](https://www.w3.org/WAI/ARIA/apg/patterns/treeview/) would be a better fit for marking up this component.

- Should file trees be a `<figure>` and require a `<figcaption>` to ensure there is a human-authored summary of what a file tree shows?

**Comment by Yan-Thomas** [comment]
> Investigate whether [the Tree View Pattern documented in the ARIA Authoring Practices Guide](https://www.w3.org/WAI/ARIA/apg/patterns/treeview/) would be a better fit for marking up this component.

Yes, it gives a few advantages for us:
- Descriptive role, so NVDA for example, reads it out as "(aria-label) TreeView". Think something like "Initial Astro project filesystem TreeView", in which people navigating using Tab will know exactly what the TreeView is talking about without relying on the surrounding text.

- It allows the user to navigate the entire TreeView (not only directories) using only the keyboard. Currently, we can't focus on each individual file, but with the TreeView using the arrow keys it's possible, which is a better UX than using "read new line" shortcuts in screen readers.

- With all the internal TreeView navigation happening with arrow keys, the user can easily Tab inside/outside the component (so, no need to tab through all the directories inside a TreeView to be able to leave it and go to the next interactive element)

- Clear file depth (will be especially useful if we end up having a use-case for a TreeView with many directories of different depths)

> Should file trees be a `<figure>` and require a `<figcaption>` to ensure there is a human-authored summary of what a file tree shows?

I don't have strong reasons against doing so. I tested this locally and it worked well. Though, I'm not exactly sure what would we summarize in it. Would that be the main files in it? (How could we judge what are the most important?) The context of the FileTree? (e.g "Content Collections folder TreeView" - in this case, the aria-label in the TreeView is enough).

I searched through a few UI Libraries with this component, like [MUI](https://mui.com/material-ui/react-tree-view/) and none of them used `<figure>`. It feels redundant to me, even more considering we tend to explain the context of the FileTree in the paragraph above/below it, which has a similar effect to adding a figcaption.

**TL;DR:** About using APG's recommended practices, go for it, it has clear advantages. Figure/figcaption are perhaps redundant.

**Comment by delucis** [comment]
Awesome work, @Yan-Thomas! Really appreciate this legwork trying stuff out. The one piece I’m still curious about is if it would be weird to have a TreeView in the middle of content, as _part_ of content. The APG seems mostly aimed at “I need to build a file tree in my native-like app that browses files” rather than our case.

But thanks to your help, I do think I’ll give it a try. I want to make this component part of our shareable docs component library for other people to use and having a pretty, accessible file tree seems like a huge win over existing tools.

**Comment by Jutanium** [comment]
@crutchcorn said he might have time for this one!

**Comment by Jothsa** [comment]
I might be able to give this a shot. It looks like to comply with the [ARIA Guidelines](https://www.w3.org/WAI/ARIA/apg/patterns/treeview/examples/treeview-1a/#rps_label:~:text=aria%2Dlabelledby%3D%22ID_REFERENCE,the%20tree.), the trees will need to have a label which will require changing all the existing usages of the tree component. In [some cases](https://docs.astro.build/en/core-concepts/project-structure/#example-project-tree), there's already a nearby header that will easily work, but in [others](https://docs.astro.build/en/guides/upgrade-to/v2/#changed-_astro-folder-for-build-assets), it's not so clear.

**Comment by delucis** [comment]
Closing as this component is moving to [the Starlight repo](https://github.com/withastro/starlight), although we do still need to do this work there.
